himbaechel/uarch/gatemate/pack_clocking.cc

```diff
@@ -682,7 +682,116 @@ void GateMatePacker::copy_clocks()
+                    // Connect IOSEL and CPE_IBUF
+                    gpio_new->connectPorts(id_Y, iosel_new, id_GPIO_IN);
+                }
+                uarch->clkin[new_die]->disconnectPort(ctx->idf("CLK%d", i));
+                if (iosel_new->getPort(id_IN1))
+                    uarch->clkin[new_die]->connectPort(ctx->idf("CLK%d", i), iosel_new->getPort(id_IN1));
+                else
+                    iosel_new->connectPorts(id_IN1, uarch->clkin[new_die], ctx->idf("CLK%d", i));
+            }
+        }
+    }
+}
+
+void GateMatePacker::strategy_clk1()
+{
+    log_info("Reuse CLK1 for clock distribution..\n");
+    NetInfo *net = uarch->glbout[0]->getPort(id_GLB0);
+    NetInfo *new_clk1 = ctx->createNet(ctx->id("$clk1$pin"));
+    for (int new_die = 0; new_die < uarch->dies; new_die++) {
+        CellInfo *iosel = create_cell_ptr(id_IOSEL, ctx->idf("$iosel_clk1$die%d", new_die));
+        iosel->setParam(id_DELAY_IBF, Property(1, 16));
+        iosel->setParam(id_INPUT_ENABLE, Property(1, 1));
+        if (new_die == 0) {
+            // On die 0 it should be output as well
+            iosel->setParam(id_DELAY_OBF, Property(1, 16));
+            iosel->setParam(id_OE_ENABLE, Property(1, 1));
+            iosel->setParam(id_OUT_SIGNAL, Property(1, 1));
+            iosel->setParam(id_SLEW, Property(1, 1));
+        }
+
+        BelId bel = ctx->getBelByLocation(uarch->locations[std::make_pair(ctx->id("IO_SB_A7"), new_die)]);
+        ctx->bindBel(bel, iosel, PlaceStrength::STRENGTH_FIXED);
+
+        CellInfo *gpio = create_cell_ptr((new_die ? id_CPE_IBUF : id_CPE_IOBUF), ctx->idf("$clk1$die%d", new_die));
+        Loc loc = ctx->getBelLocation(bel);
+        ctx->bindBel(ctx->getBelByLocation({loc.x, loc.y, 0}), gpio, PlaceStrength::STRENGTH_FIXED);
+
+        uarch->clkin[new_die]->connectPort(id_CLK1, new_clk1);
+        uarch->clkin[new_die]->params[ctx->id("REF1")] = Property(1, 3);
+        uarch->glbout[new_die]->params[ctx->id("GLB1_EN")] = Property(Property::State::S1);
+        uarch->glbout[new_die]->params[ctx->id("GLB1_CFG")] = Property(0, 3);
+        uarch->clkin[new_die]->connectPorts(ctx->id("CLK_REF1"), uarch->glbout[new_die], ctx->id("CLK_REF_OUT1"));
+
+        gpio->connectPorts(id_Y, iosel, id_GPIO_IN);
+
+        if (new_die == 0) {
+            iosel->connectPort(id_OUT1, ctx->getNetByAlias(uarch->global_signals.begin()->first->name));
+            CellInfo *cpe = move_ram_o_fixed(iosel, id_OUT1, loc).first;
+            uarch->ignore.emplace(cpe->name);
+
+            iosel->connectPorts(id_GPIO_OUT, gpio, id_A);
+            iosel->connectPorts(id_GPIO_EN, gpio, id_T);
+            gpio->connectPort(id_IO, new_clk1);
+        } else
+            gpio->connectPort(id_I, new_clk1);
+
+        NetInfo *new_signal = ctx->createNet(ctx->idf("%s$die%d", net->name.c_str(ctx), new_die));
+        uarch->glbout[new_die]->connectPort(ctx->id("GLB1"), new_signal);
+        copy_constraint(net, new_signal);
```

Comment:
note to myself as much as anything else - we really need support for clock constraints with phase shifts for timing analysis to work in this kind of situation properly